plugin/mlsdbonlinelocator.cpp

```diff
@@ -88,8 +111,88 @@ void MlsdbOnlineLocator::defaultVoiceModemChanged(const QString &modem)
+        const bool intervalExceeded = oldQuery.first.isNull() || oldQuery.first.msecsTo(currDt) >= m_adaptiveInterval;
+        const bool moreInfo = map.keys().size() > oldQuery.second.keys().size();
+        const bool newCells = cellIdsFromQueryData(oldQuery.second) != cellIdsFromQueryData(map);
+
+        if (firstTimeQuery || intervalExceeded || moreInfo || newCells) {
+            // adaptively back-off future requests to avoid server-side throttling.
+            static quint32 backOffFactor = 8;
+
+            // we want to aim for approximately 6 minutes per request.
+            const qint64 deltaMsecs = (m_queryTimestamps.size() < 3)
+                                    ? 0
+                                    : (m_queryTimestamps.first() - m_queryTimestamps.last());
+            const double minutesPerQuery = (m_queryTimestamps.size() < 3)
+                                         ? 6
+                                         : ((deltaMsecs / (1000.0 * 60.0)) / m_queryTimestamps.size());
+            if (minutesPerQuery > 6 || backOffFactor > 64) {
+                // it's been a long time since the last request, reduce the back off factor.
+                backOffFactor = backOffFactor <= 2 ? 1 : (backOffFactor / 2);
+            } else if (minutesPerQuery < 4) {
+                // too many recent requests, increase the back-off factor.
+                backOffFactor = backOffFactor >= 32 ? 64 : (backOffFactor * 2);
+            }
+
+            // max interval will be about 12 minutes (1 + 10.667 minutes).
+            m_adaptiveInterval = REQUEST_BASE_ADAPTIVE_INTERVAL + (REQUEST_MODIFY_ADAPTIVE_INTERVAL * backOffFactor);
+
+            if (backOffFactor == 1 || intervalExceeded) {
+                // return the query data for the request.
+                qCDebug(lcGeoclueMlsdbOnline) << "Performing MLS online query due to conditions:"
+                                              << "first:" << firstTimeQuery
+                                              << "interval:" << intervalExceeded
+                                              << "info:" << moreInfo
+                                              << "cells:" << newCells;
+                m_queryTimestamps.prepend(QDateTime::currentMSecsSinceEpoch());
+                if (m_queryTimestamps.size() > REQUEST_TIMESTAMPS_TO_TRACK) {
+                    m_queryTimestamps.removeLast();
+                }
+                return qMakePair(currDt, map);
+            } else {
+                qCDebug(lcGeoclueMlsdbOnline) << "Locally throttling online MLS query due to interval";
+            }
+        } else {
+            qCDebug(lcGeoclueMlsdbOnline) << "No required conditions true for online MLS query!";
+        }
+    }
+
+    return qMakePair(QDateTime(), QVariantMap());
+}
+
+bool MlsdbOnlineLocator::findLocation(const QPair<QDateTime, QVariantMap> &query)
 {
+    if (query.second.isEmpty()) {
```

Comment:
Currently we do not ever endup in this as you check the value before calling findLocation as well. Maybe just call this and check also the time inside this?

Reply:
Yes, will do